core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushdownFilterIntoRowNumber.java

```diff
@@ -93,32 +93,10 @@ public Result apply(FilterNode node, Captures captures, Context context)
             return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
         }
 
-        boolean needRewriteSource = true;
-        if (source.getMaxRowCountPerPartition().isPresent()) {
-            int newRowCountPerPartition = Math.min(source.getMaxRowCountPerPartition().get(), upperBound.getAsInt());
-            // No need to rewrite only source node if rowCountPerPartition results in identical.
-            needRewriteSource = source.getMaxRowCountPerPartition().get() != newRowCountPerPartition;
-            if (needRewriteSource) {
-                source = new RowNumberNode(
-                        source.getId(),
-                        source.getSource(),
-                        source.getPartitionBy(),
-                        source.isOrderSensitive(),
-                        source.getRowNumberSymbol(),
-                        Optional.of(newRowCountPerPartition),
-                        source.getHashSymbol());
-            }
-        }
-        else {
-            // Pushdown the upper bound of the filter
-            source = new RowNumberNode(
-                    source.getId(),
-                    source.getSource(),
-                    source.getPartitionBy(),
-                    source.isOrderSensitive(),
-                    source.getRowNumberSymbol(),
-                    Optional.of(upperBound.getAsInt()),
-                    source.getHashSymbol());
+        RowNumberNode merged = mergeLimit(source, upperBound.getAsInt());
+        boolean needRewriteSource = !merged.getMaxRowCountPerPartition().equals(source.getMaxRowCountPerPartition());
+        if (needRewriteSource) {
```

Comment:
If you want, you can probably just inline this boolean variable directly into the if-condition. But just a personal preference thing (not a big deal if not).

Reply:
Thanks. Will do.

Reply:
Sorry. `needRewriteSource` is also used in L.103. It cannot be inlined.